src/transformers/pipelines/fill_in_middle.py

```diff
@@ -0,0 +1,368 @@
+        forward_params = generate_kwargs
+
+        postprocess_params = {}
+
+        if return_full_text is not None and return_type is None:
+            if return_text is not None:
+                raise ValueError("`return_text` is mutually exclusive with `return_full_text`")
+            if return_tensors is not None:
+                raise ValueError("`return_full_text` is mutually exclusive with `return_tensors`")
+            return_type = ReturnType.FULL_TEXT if return_full_text else ReturnType.NEW_TEXT
+        if return_tensors is not None and return_type is None:
+            if return_text is not None:
+                raise ValueError("`return_text` is mutually exclusive with `return_tensors`")
+            return_type = ReturnType.TENSORS
+        if return_type is not None:
+            postprocess_params["return_type"] = return_type
+        if clean_up_tokenization_spaces is not None:
+            postprocess_params["clean_up_tokenization_spaces"] = clean_up_tokenization_spaces
+
+        if stop_sequence is not None:
+            stop_sequence_ids = self.tokenizer.encode(stop_sequence, add_special_tokens=False)
+            if len(stop_sequence_ids) > 1:
+                warnings.warn(
+                    "Stopping on a multiple token sequence is not yet supported on transformers. The first token of"
+                    " the stop sequence will be used as the stop sequence string in the interim."
+                )
+            generate_kwargs["eos_token_id"] = stop_sequence_ids[0]
+
+        return preprocess_params, forward_params, postprocess_params
+
+    def preprocess(
+        self,
+        prompt_text,
+        infill_token,
+        mode,
+        prefix="",
+        handle_long_generation=None,
+        add_special_tokens=False,
+        **generate_kwargs,
+    ):
+        # Ensure the prompt_text contains the infill token
+        self.ensure_infill_token(prompt_text, infill_token)
+
+        # Assemble the inputs in the PSM (Prefix-Suffix-Middle) format based on the model architecture
+        # Currently, codellama model variants support using only the "<FILL_ME>" token, so we won't overcomplicate things
+        if "codellama" not in self.model.name_or_path.lower():
+            # Extract prefix and suffix
+            input_prefix, input_suffix = self.extract_prefix_suffix(prompt_text, infill_token)
+
+            if mode == "psm":
+                prompt_text = (
+                    self.DEFAULT_PREFIX_TOKEN
+                    + input_prefix
+                    + self.DEFAULT_SUFFIX_TOKEN
+                    + input_suffix
+                    + self.DEFAULT_MIDDLE_TOKEN
+                )
+            else:
+                prompt_text = (
+                    self.DEFAULT_SUFFIX_TOKEN
+                    + input_suffix
+                    + self.DEFAULT_PREFIX_TOKEN
+                    + input_prefix
+                    + self.DEFAULT_MIDDLE_TOKEN
+                )
```

Comment:
not sure I understand why we only support Llama for this case?

Reply:
Of the two models (that I am aware of) that support the FIM objective, CodeLlama has the support of directly using the `<FILL_ME>` token whereas starcoder doesn't. We have to manually add the prefix, suffix and middle tokens, hence the check.

Reply:
Yes, but unifying this would be better! But got it for this now worries